## File-1

**Target** : Design a baseline models and simple training/test scripts that has less than 10000 parameters and shows an accuracy of atleast 99.2-99.3% once in 15 epochs. Choose the model/models  to proceed further with.

**Result**
- All of the 6 models designed hit the 99.2% mark, but none of them reached 99.3.
- Net1(7922 parameters) and Net4(9972 parameters) were the best models with highest test accuracy as  - 99.27 and 99.24 respectively and highest train accuracy as 99.47 and 99.5 respectively

**Analysis**
- Net  1 and 4 are the best models because they showed high training accuracy(around 99.4-99.5) in the first 15 epochs.
- Although models 1 and 4 are overfitting, by using regularizations in the next step, I can increase the test accuracy. These 2 models were chosen for further steps because they **learn better** than the other 4.

In [ ]:
!pip install torchsummary
import torch
import torch.optim as optim
from torchsummary import summary
from train import train
from test import test
from dataloader import train_loader,test_loader
from models import Net1,Net2,Net3,Net4,Net5,Net6

In [ ]:
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")

## Net 1 training and analysis
### Results
- Best training accuracy  - 99.47
- Best test accuracy - 99.27
## Analysis 
- Model does show overfitting after the 8th epoch(Significant differnnece between the training and test accuracy can be seen)
- Accuaracy fluctuates around 99.2%. LR might be too high at this stage of training.

In [ ]:
model = Net1().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
for epoch in range(1, 21):
    print(f"Epoch {epoch}")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

Epoch 1


  0%|          | 0/469 [00:00<?, ?it/s]/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:718: UserWarning: Named tensors and all their associated APIs are an experimental feature and subject to change. Please do not use them for anything important until they are released as stable. (Triggered internally at  /pytorch/c10/core/TensorImpl.h:1156.)
  return torch.max_pool2d(input, kernel_size, stride, padding, dilation, ceil_mode)
Loss=0.08884599059820175 Batch_id=468 Train Accuracy=91.05: 100%|██████████| 469/469 [00:20<00:00, 22.76it/s]



Test set: Average loss: 0.0793, Test Accuracy: 9805/10000 (98.05%)

Epoch 2


Loss=0.0872727632522583 Batch_id=468 Train Accuracy=98.22: 100%|██████████| 469/469 [00:20<00:00, 23.01it/s]



Test set: Average loss: 0.0430, Test Accuracy: 9886/10000 (98.86%)

Epoch 3


Loss=0.01676887273788452 Batch_id=468 Train Accuracy=98.65: 100%|██████████| 469/469 [00:20<00:00, 22.83it/s]



Test set: Average loss: 0.0423, Test Accuracy: 9881/10000 (98.81%)

Epoch 4


Loss=0.07911486178636551 Batch_id=468 Train Accuracy=98.74: 100%|██████████| 469/469 [00:20<00:00, 22.92it/s]



Test set: Average loss: 0.0390, Test Accuracy: 9880/10000 (98.80%)

Epoch 5


Loss=0.018319858238101006 Batch_id=468 Train Accuracy=98.89: 100%|██████████| 469/469 [00:20<00:00, 23.03it/s]



Test set: Average loss: 0.0334, Test Accuracy: 9902/10000 (99.02%)

Epoch 6


Loss=0.0072396546602249146 Batch_id=468 Train Accuracy=99.05: 100%|██████████| 469/469 [00:20<00:00, 22.96it/s]



Test set: Average loss: 0.0367, Test Accuracy: 9880/10000 (98.80%)

Epoch 7


Loss=0.010884511284530163 Batch_id=468 Train Accuracy=99.10: 100%|██████████| 469/469 [00:20<00:00, 22.83it/s]



Test set: Average loss: 0.0301, Test Accuracy: 9910/10000 (99.10%)

Epoch 8


Loss=0.04698944091796875 Batch_id=468 Train Accuracy=99.19: 100%|██████████| 469/469 [00:20<00:00, 22.96it/s]



Test set: Average loss: 0.0334, Test Accuracy: 9903/10000 (99.03%)

Epoch 9


Loss=0.009081305004656315 Batch_id=468 Train Accuracy=99.26: 100%|██████████| 469/469 [00:20<00:00, 22.76it/s]



Test set: Average loss: 0.0281, Test Accuracy: 9920/10000 (99.20%)

Epoch 10


Loss=0.012733064591884613 Batch_id=468 Train Accuracy=99.34: 100%|██████████| 469/469 [00:20<00:00, 22.82it/s]



Test set: Average loss: 0.0304, Test Accuracy: 9901/10000 (99.01%)

Epoch 11


Loss=0.004559930879622698 Batch_id=468 Train Accuracy=99.31: 100%|██████████| 469/469 [00:20<00:00, 22.84it/s]



Test set: Average loss: 0.0236, Test Accuracy: 9922/10000 (99.22%)

Epoch 12


Loss=0.019102534279227257 Batch_id=468 Train Accuracy=99.34: 100%|██████████| 469/469 [00:20<00:00, 22.76it/s]



Test set: Average loss: 0.0271, Test Accuracy: 9909/10000 (99.09%)

Epoch 13


Loss=0.006805206183344126 Batch_id=468 Train Accuracy=99.40: 100%|██████████| 469/469 [00:20<00:00, 22.74it/s]



Test set: Average loss: 0.0245, Test Accuracy: 9919/10000 (99.19%)

Epoch 14


Loss=0.00600385433062911 Batch_id=468 Train Accuracy=99.40: 100%|██████████| 469/469 [00:20<00:00, 22.82it/s]



Test set: Average loss: 0.0262, Test Accuracy: 9913/10000 (99.13%)

Epoch 15


Loss=0.004877082537859678 Batch_id=468 Train Accuracy=99.47: 100%|██████████| 469/469 [00:20<00:00, 22.74it/s]



Test set: Average loss: 0.0243, Test Accuracy: 9927/10000 (99.27%)

Epoch 16


Loss=0.014053675346076488 Batch_id=468 Train Accuracy=99.44: 100%|██████████| 469/469 [00:20<00:00, 22.55it/s]



Test set: Average loss: 0.0225, Test Accuracy: 9920/10000 (99.20%)

Epoch 17


Loss=0.007912642322480679 Batch_id=468 Train Accuracy=99.52: 100%|██████████| 469/469 [00:20<00:00, 22.47it/s]



Test set: Average loss: 0.0279, Test Accuracy: 9914/10000 (99.14%)

Epoch 18


Loss=0.009956047870218754 Batch_id=468 Train Accuracy=99.53: 100%|██████████| 469/469 [00:20<00:00, 22.51it/s]



Test set: Average loss: 0.0226, Test Accuracy: 9922/10000 (99.22%)

Epoch 19


Loss=0.01595579832792282 Batch_id=468 Train Accuracy=99.55: 100%|██████████| 469/469 [00:20<00:00, 22.86it/s]



Test set: Average loss: 0.0253, Test Accuracy: 9916/10000 (99.16%)

Epoch 20


Loss=0.009689251892268658 Batch_id=468 Train Accuracy=99.58: 100%|██████████| 469/469 [00:20<00:00, 22.76it/s]



Test set: Average loss: 0.0230, Test Accuracy: 9929/10000 (99.29%)



## Net 2 training and analysis
### Results
- Best training accuracy  - 99.31
- Best test accuracy - 99.06
## Analysis 
- Model isnt able to learn very well as the tes and training accuracy are quite low. Prefferably drop this model and wont include it in further analysis

In [ ]:
model = Net2().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
for epoch in range(1, 21):
    print(f"Epoch {epoch}")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

Epoch 1


Loss=0.18017078936100006 Batch_id=468 Train Accuracy=87.16: 100%|██████████| 469/469 [00:20<00:00, 22.73it/s]



Test set: Average loss: 0.1206, Test Accuracy: 9707/10000 (97.07%)

Epoch 2


Loss=0.06091884896159172 Batch_id=468 Train Accuracy=97.79: 100%|██████████| 469/469 [00:20<00:00, 22.62it/s]



Test set: Average loss: 0.0850, Test Accuracy: 9752/10000 (97.52%)

Epoch 3


Loss=0.03958681598305702 Batch_id=468 Train Accuracy=98.23: 100%|██████████| 469/469 [00:20<00:00, 22.58it/s]



Test set: Average loss: 0.0501, Test Accuracy: 9847/10000 (98.47%)

Epoch 4


Loss=0.024092065170407295 Batch_id=468 Train Accuracy=98.54: 100%|██████████| 469/469 [00:20<00:00, 22.62it/s]



Test set: Average loss: 0.0439, Test Accuracy: 9865/10000 (98.65%)

Epoch 5


Loss=0.1019582524895668 Batch_id=468 Train Accuracy=98.73: 100%|██████████| 469/469 [00:20<00:00, 22.79it/s]



Test set: Average loss: 0.0484, Test Accuracy: 9847/10000 (98.47%)

Epoch 6


Loss=0.03188376501202583 Batch_id=468 Train Accuracy=98.90: 100%|██████████| 469/469 [00:20<00:00, 22.89it/s]



Test set: Average loss: 0.0475, Test Accuracy: 9851/10000 (98.51%)

Epoch 7


Loss=0.029429323971271515 Batch_id=468 Train Accuracy=98.95: 100%|██████████| 469/469 [00:20<00:00, 22.59it/s]



Test set: Average loss: 0.0354, Test Accuracy: 9882/10000 (98.82%)

Epoch 8


Loss=0.049206748604774475 Batch_id=468 Train Accuracy=99.00: 100%|██████████| 469/469 [00:20<00:00, 22.58it/s]



Test set: Average loss: 0.0326, Test Accuracy: 9906/10000 (99.06%)

Epoch 9


Loss=0.06280835717916489 Batch_id=468 Train Accuracy=99.09: 100%|██████████| 469/469 [00:20<00:00, 22.59it/s]



Test set: Average loss: 0.0371, Test Accuracy: 9881/10000 (98.81%)

Epoch 10


Loss=0.008027918636798859 Batch_id=468 Train Accuracy=99.10: 100%|██████████| 469/469 [00:20<00:00, 22.52it/s]



Test set: Average loss: 0.0331, Test Accuracy: 9906/10000 (99.06%)

Epoch 11


Loss=0.03537142649292946 Batch_id=468 Train Accuracy=99.20: 100%|██████████| 469/469 [00:20<00:00, 22.57it/s]



Test set: Average loss: 0.0393, Test Accuracy: 9875/10000 (98.75%)

Epoch 12


Loss=0.010657480917870998 Batch_id=468 Train Accuracy=99.14: 100%|██████████| 469/469 [00:20<00:00, 22.73it/s]



Test set: Average loss: 0.0401, Test Accuracy: 9870/10000 (98.70%)

Epoch 13


Loss=0.0046744998544454575 Batch_id=468 Train Accuracy=99.28: 100%|██████████| 469/469 [00:20<00:00, 22.45it/s]



Test set: Average loss: 0.0339, Test Accuracy: 9887/10000 (98.87%)

Epoch 14


Loss=0.022248072549700737 Batch_id=468 Train Accuracy=99.23: 100%|██████████| 469/469 [00:20<00:00, 22.52it/s]



Test set: Average loss: 0.0335, Test Accuracy: 9885/10000 (98.85%)

Epoch 15


Loss=0.02528669871389866 Batch_id=468 Train Accuracy=99.31: 100%|██████████| 469/469 [00:20<00:00, 22.56it/s]



Test set: Average loss: 0.0305, Test Accuracy: 9900/10000 (99.00%)

Epoch 16


Loss=0.012467061169445515 Batch_id=468 Train Accuracy=99.29: 100%|██████████| 469/469 [00:20<00:00, 22.56it/s]



Test set: Average loss: 0.0308, Test Accuracy: 9896/10000 (98.96%)

Epoch 17


Loss=0.07174592465162277 Batch_id=468 Train Accuracy=99.39: 100%|██████████| 469/469 [00:20<00:00, 22.43it/s]



Test set: Average loss: 0.0372, Test Accuracy: 9877/10000 (98.77%)

Epoch 18


Loss=0.021925488486886024 Batch_id=468 Train Accuracy=99.39: 100%|██████████| 469/469 [00:20<00:00, 22.45it/s]



Test set: Average loss: 0.0344, Test Accuracy: 9886/10000 (98.86%)

Epoch 19


Loss=0.007976595312356949 Batch_id=468 Train Accuracy=99.34: 100%|██████████| 469/469 [00:20<00:00, 22.57it/s]



Test set: Average loss: 0.0258, Test Accuracy: 9922/10000 (99.22%)

Epoch 20


Loss=0.020487459376454353 Batch_id=468 Train Accuracy=99.38: 100%|██████████| 469/469 [00:20<00:00, 22.39it/s]



Test set: Average loss: 0.0253, Test Accuracy: 9918/10000 (99.18%)



## Net 3 training and analysis
### Results
- Best training accuracy  - 99.03
- Best test accuracy - 99.13
## Analysis 
- Model does show underfitting even after 15 epochs. Maybe thats because I havent added relu after every convlutional layer and instead added it at the end of each block only.

In [ ]:
model = Net3().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
for epoch in range(1, 21):
    print(f"Epoch {epoch}")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

Epoch 1


Loss=0.19713521003723145 Batch_id=468 Train Accuracy=90.26: 100%|██████████| 469/469 [00:20<00:00, 22.79it/s]



Test set: Average loss: 0.0941, Test Accuracy: 9748/10000 (97.48%)

Epoch 2


Loss=0.06190653145313263 Batch_id=468 Train Accuracy=97.59: 100%|██████████| 469/469 [00:20<00:00, 22.89it/s]



Test set: Average loss: 0.0639, Test Accuracy: 9842/10000 (98.42%)

Epoch 3


Loss=0.03327406942844391 Batch_id=468 Train Accuracy=98.10: 100%|██████████| 469/469 [00:20<00:00, 22.64it/s]



Test set: Average loss: 0.0520, Test Accuracy: 9841/10000 (98.41%)

Epoch 4


Loss=0.06273108720779419 Batch_id=468 Train Accuracy=98.36: 100%|██████████| 469/469 [00:20<00:00, 22.90it/s]



Test set: Average loss: 0.0480, Test Accuracy: 9862/10000 (98.62%)

Epoch 5


Loss=0.03930334746837616 Batch_id=468 Train Accuracy=98.59: 100%|██████████| 469/469 [00:20<00:00, 22.91it/s]



Test set: Average loss: 0.0359, Test Accuracy: 9890/10000 (98.90%)

Epoch 6


Loss=0.03304987773299217 Batch_id=468 Train Accuracy=98.64: 100%|██████████| 469/469 [00:20<00:00, 22.84it/s]



Test set: Average loss: 0.0408, Test Accuracy: 9879/10000 (98.79%)

Epoch 7


Loss=0.09777174144983292 Batch_id=468 Train Accuracy=98.72: 100%|██████████| 469/469 [00:20<00:00, 23.09it/s]



Test set: Average loss: 0.0356, Test Accuracy: 9900/10000 (99.00%)

Epoch 8


Loss=0.023584887385368347 Batch_id=468 Train Accuracy=98.81: 100%|██████████| 469/469 [00:20<00:00, 23.10it/s]



Test set: Average loss: 0.0411, Test Accuracy: 9863/10000 (98.63%)

Epoch 9


Loss=0.04146165773272514 Batch_id=468 Train Accuracy=98.86: 100%|██████████| 469/469 [00:20<00:00, 23.18it/s]



Test set: Average loss: 0.0350, Test Accuracy: 9892/10000 (98.92%)

Epoch 10


Loss=0.03685365989804268 Batch_id=468 Train Accuracy=98.86: 100%|██████████| 469/469 [00:20<00:00, 23.00it/s]



Test set: Average loss: 0.0428, Test Accuracy: 9860/10000 (98.60%)

Epoch 11


Loss=0.012781561352312565 Batch_id=468 Train Accuracy=98.97: 100%|██████████| 469/469 [00:20<00:00, 22.96it/s]



Test set: Average loss: 0.0277, Test Accuracy: 9920/10000 (99.20%)

Epoch 12


Loss=0.032579172402620316 Batch_id=468 Train Accuracy=99.00: 100%|██████████| 469/469 [00:20<00:00, 22.96it/s]



Test set: Average loss: 0.0292, Test Accuracy: 9913/10000 (99.13%)

Epoch 13


Loss=0.050041746348142624 Batch_id=468 Train Accuracy=99.05: 100%|██████████| 469/469 [00:20<00:00, 23.12it/s]



Test set: Average loss: 0.0283, Test Accuracy: 9910/10000 (99.10%)

Epoch 14


Loss=0.03774504363536835 Batch_id=468 Train Accuracy=99.07: 100%|██████████| 469/469 [00:20<00:00, 22.54it/s]



Test set: Average loss: 0.0313, Test Accuracy: 9908/10000 (99.08%)

Epoch 15


Loss=0.01972408965229988 Batch_id=468 Train Accuracy=99.08: 100%|██████████| 469/469 [00:20<00:00, 23.02it/s]



Test set: Average loss: 0.0295, Test Accuracy: 9913/10000 (99.13%)

Epoch 16


Loss=0.021640196442604065 Batch_id=468 Train Accuracy=99.11: 100%|██████████| 469/469 [00:20<00:00, 22.97it/s]



Test set: Average loss: 0.0286, Test Accuracy: 9910/10000 (99.10%)

Epoch 17


Loss=0.030169397592544556 Batch_id=468 Train Accuracy=99.06: 100%|██████████| 469/469 [00:20<00:00, 22.52it/s]



Test set: Average loss: 0.0269, Test Accuracy: 9908/10000 (99.08%)

Epoch 18


Loss=0.018717577680945396 Batch_id=468 Train Accuracy=99.13: 100%|██████████| 469/469 [00:20<00:00, 22.88it/s]



Test set: Average loss: 0.0311, Test Accuracy: 9898/10000 (98.98%)

Epoch 19


Loss=0.010875135660171509 Batch_id=468 Train Accuracy=99.14: 100%|██████████| 469/469 [00:20<00:00, 22.40it/s]



Test set: Average loss: 0.0277, Test Accuracy: 9910/10000 (99.10%)

Epoch 20


Loss=0.033334746956825256 Batch_id=468 Train Accuracy=99.22: 100%|██████████| 469/469 [00:20<00:00, 22.42it/s]



Test set: Average loss: 0.0266, Test Accuracy: 9914/10000 (99.14%)



## Net 4 training and analysis
### Results
- Best training accuracy  - 99.58
- Best test accuracy - 99.24
## Analysis 
- Model does show overfitting after the 8th epoch(Significant differnnece between the training and test accuracy can be seen)
- Accuaracy fluctuates around 99.2%. LR might be too high at this stage of training.

In [ ]:
model = Net4().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
for epoch in range(1, 21):
    print(f"Epoch {epoch}")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

Epoch 1


Loss=0.12577272951602936 Batch_id=468 Train Accuracy=86.52: 100%|██████████| 469/469 [00:20<00:00, 22.69it/s]



Test set: Average loss: 0.0953, Test Accuracy: 9759/10000 (97.59%)

Epoch 2


Loss=0.1225116029381752 Batch_id=468 Train Accuracy=97.95: 100%|██████████| 469/469 [00:20<00:00, 22.94it/s]



Test set: Average loss: 0.0497, Test Accuracy: 9870/10000 (98.70%)

Epoch 3


Loss=0.014031368307769299 Batch_id=468 Train Accuracy=98.47: 100%|██████████| 469/469 [00:20<00:00, 22.85it/s]



Test set: Average loss: 0.0433, Test Accuracy: 9876/10000 (98.76%)

Epoch 4


Loss=0.010658681392669678 Batch_id=468 Train Accuracy=98.78: 100%|██████████| 469/469 [00:20<00:00, 22.70it/s]



Test set: Average loss: 0.0354, Test Accuracy: 9887/10000 (98.87%)

Epoch 5


Loss=0.008672471158206463 Batch_id=468 Train Accuracy=98.94: 100%|██████████| 469/469 [00:20<00:00, 22.82it/s]



Test set: Average loss: 0.0375, Test Accuracy: 9881/10000 (98.81%)

Epoch 6


Loss=0.0658513531088829 Batch_id=468 Train Accuracy=99.07: 100%|██████████| 469/469 [00:20<00:00, 22.83it/s]



Test set: Average loss: 0.0320, Test Accuracy: 9901/10000 (99.01%)

Epoch 7


Loss=0.09191287308931351 Batch_id=468 Train Accuracy=99.17: 100%|██████████| 469/469 [00:20<00:00, 23.01it/s]



Test set: Average loss: 0.0312, Test Accuracy: 9903/10000 (99.03%)

Epoch 8


Loss=0.03416457399725914 Batch_id=468 Train Accuracy=99.22: 100%|██████████| 469/469 [00:20<00:00, 22.84it/s]



Test set: Average loss: 0.0272, Test Accuracy: 9909/10000 (99.09%)

Epoch 9


Loss=0.03005487658083439 Batch_id=468 Train Accuracy=99.28: 100%|██████████| 469/469 [00:20<00:00, 22.85it/s]



Test set: Average loss: 0.0247, Test Accuracy: 9916/10000 (99.16%)

Epoch 10


Loss=0.04818034544587135 Batch_id=468 Train Accuracy=99.28: 100%|██████████| 469/469 [00:20<00:00, 22.42it/s]



Test set: Average loss: 0.0247, Test Accuracy: 9920/10000 (99.20%)

Epoch 11


Loss=0.007307508494704962 Batch_id=468 Train Accuracy=99.34: 100%|██████████| 469/469 [00:20<00:00, 22.63it/s]



Test set: Average loss: 0.0365, Test Accuracy: 9889/10000 (98.89%)

Epoch 12


Loss=0.005351261701434851 Batch_id=468 Train Accuracy=99.42: 100%|██████████| 469/469 [00:20<00:00, 22.69it/s]



Test set: Average loss: 0.0277, Test Accuracy: 9919/10000 (99.19%)

Epoch 13


Loss=0.04044400528073311 Batch_id=468 Train Accuracy=99.40: 100%|██████████| 469/469 [00:20<00:00, 22.59it/s]



Test set: Average loss: 0.0267, Test Accuracy: 9924/10000 (99.24%)

Epoch 14


Loss=0.0044978405348956585 Batch_id=468 Train Accuracy=99.43: 100%|██████████| 469/469 [00:20<00:00, 22.66it/s]



Test set: Average loss: 0.0269, Test Accuracy: 9915/10000 (99.15%)

Epoch 15


Loss=0.03223394975066185 Batch_id=468 Train Accuracy=99.50: 100%|██████████| 469/469 [00:20<00:00, 22.81it/s]



Test set: Average loss: 0.0266, Test Accuracy: 9924/10000 (99.24%)

Epoch 16


Loss=0.02328270673751831 Batch_id=468 Train Accuracy=99.53: 100%|██████████| 469/469 [00:20<00:00, 22.44it/s]



Test set: Average loss: 0.0282, Test Accuracy: 9906/10000 (99.06%)

Epoch 17


Loss=0.00824587233364582 Batch_id=468 Train Accuracy=99.55: 100%|██████████| 469/469 [00:20<00:00, 22.42it/s]



Test set: Average loss: 0.0252, Test Accuracy: 9921/10000 (99.21%)

Epoch 18


Loss=0.0027958350256085396 Batch_id=468 Train Accuracy=99.58: 100%|██████████| 469/469 [00:20<00:00, 22.71it/s]



Test set: Average loss: 0.0244, Test Accuracy: 9929/10000 (99.29%)

Epoch 19


Loss=0.025197545066475868 Batch_id=468 Train Accuracy=99.58: 100%|██████████| 469/469 [00:20<00:00, 22.55it/s]



Test set: Average loss: 0.0259, Test Accuracy: 9920/10000 (99.20%)

Epoch 20


Loss=0.004350319970399141 Batch_id=468 Train Accuracy=99.54: 100%|██████████| 469/469 [00:20<00:00, 22.34it/s]



Test set: Average loss: 0.0247, Test Accuracy: 9923/10000 (99.23%)



## Net 5 training and analysis
### Results
- Best training accuracy  - 99.22
- Best test accuracy - 99.18
## Analysis 
- Model does not show overfit as much as other models
- but still does not show high enough accuracy to be included in further analysis

In [ ]:
model = Net5().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
for epoch in range(1, 21):
    print(f"Epoch {epoch}")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

Epoch 1


Loss=0.16564851999282837 Batch_id=468 Train Accuracy=91.56: 100%|██████████| 469/469 [00:20<00:00, 22.76it/s]



Test set: Average loss: 0.1068, Test Accuracy: 9782/10000 (97.82%)

Epoch 2


Loss=0.0916542112827301 Batch_id=468 Train Accuracy=97.79: 100%|██████████| 469/469 [00:20<00:00, 22.70it/s]



Test set: Average loss: 0.0676, Test Accuracy: 9831/10000 (98.31%)

Epoch 3


Loss=0.02575811557471752 Batch_id=468 Train Accuracy=98.31: 100%|██████████| 469/469 [00:20<00:00, 23.04it/s]



Test set: Average loss: 0.0575, Test Accuracy: 9860/10000 (98.60%)

Epoch 4


Loss=0.06534294039011002 Batch_id=468 Train Accuracy=98.47: 100%|██████████| 469/469 [00:20<00:00, 23.02it/s]



Test set: Average loss: 0.0441, Test Accuracy: 9884/10000 (98.84%)

Epoch 5


Loss=0.021900596097111702 Batch_id=468 Train Accuracy=98.67: 100%|██████████| 469/469 [00:20<00:00, 22.89it/s]



Test set: Average loss: 0.0413, Test Accuracy: 9893/10000 (98.93%)

Epoch 6


Loss=0.042226631194353104 Batch_id=468 Train Accuracy=98.76: 100%|██████████| 469/469 [00:20<00:00, 23.15it/s]



Test set: Average loss: 0.0393, Test Accuracy: 9893/10000 (98.93%)

Epoch 7


Loss=0.03713657706975937 Batch_id=468 Train Accuracy=98.91: 100%|██████████| 469/469 [00:20<00:00, 23.14it/s]



Test set: Average loss: 0.0358, Test Accuracy: 9896/10000 (98.96%)

Epoch 8


Loss=0.03372085094451904 Batch_id=468 Train Accuracy=99.01: 100%|██████████| 469/469 [00:20<00:00, 22.91it/s]



Test set: Average loss: 0.0316, Test Accuracy: 9907/10000 (99.07%)

Epoch 9


Loss=0.007059278432279825 Batch_id=468 Train Accuracy=99.03: 100%|██████████| 469/469 [00:20<00:00, 22.62it/s]



Test set: Average loss: 0.0360, Test Accuracy: 9893/10000 (98.93%)

Epoch 10


Loss=0.10296943783760071 Batch_id=468 Train Accuracy=99.12: 100%|██████████| 469/469 [00:21<00:00, 22.17it/s]



Test set: Average loss: 0.0311, Test Accuracy: 9906/10000 (99.06%)

Epoch 11


Loss=0.006442925427109003 Batch_id=468 Train Accuracy=99.15: 100%|██████████| 469/469 [00:21<00:00, 22.13it/s]



Test set: Average loss: 0.0295, Test Accuracy: 9918/10000 (99.18%)

Epoch 12


Loss=0.005552573129534721 Batch_id=468 Train Accuracy=99.17: 100%|██████████| 469/469 [00:20<00:00, 22.92it/s]



Test set: Average loss: 0.0303, Test Accuracy: 9915/10000 (99.15%)

Epoch 13


Loss=0.01799651049077511 Batch_id=468 Train Accuracy=99.22: 100%|██████████| 469/469 [00:21<00:00, 22.21it/s]



Test set: Average loss: 0.0322, Test Accuracy: 9898/10000 (98.98%)

Epoch 14


Loss=0.02707768976688385 Batch_id=468 Train Accuracy=99.22: 100%|██████████| 469/469 [00:21<00:00, 21.98it/s]



Test set: Average loss: 0.0293, Test Accuracy: 9914/10000 (99.14%)

Epoch 15


Loss=0.043494340032339096 Batch_id=468 Train Accuracy=99.27: 100%|██████████| 469/469 [00:20<00:00, 22.66it/s]



Test set: Average loss: 0.0319, Test Accuracy: 9904/10000 (99.04%)

Epoch 16


Loss=0.08676167577505112 Batch_id=468 Train Accuracy=99.27: 100%|██████████| 469/469 [00:21<00:00, 21.96it/s]



Test set: Average loss: 0.0243, Test Accuracy: 9920/10000 (99.20%)

Epoch 17


Loss=0.027738390490412712 Batch_id=468 Train Accuracy=99.28: 100%|██████████| 469/469 [00:20<00:00, 22.50it/s]



Test set: Average loss: 0.0278, Test Accuracy: 9912/10000 (99.12%)

Epoch 18


Loss=0.030887683853507042 Batch_id=468 Train Accuracy=99.32: 100%|██████████| 469/469 [00:20<00:00, 22.80it/s]



Test set: Average loss: 0.0301, Test Accuracy: 9903/10000 (99.03%)

Epoch 19


Loss=0.032251566648483276 Batch_id=468 Train Accuracy=99.32: 100%|██████████| 469/469 [00:20<00:00, 22.83it/s]



Test set: Average loss: 0.0251, Test Accuracy: 9923/10000 (99.23%)

Epoch 20


Loss=0.02297535352408886 Batch_id=468 Train Accuracy=99.34: 100%|██████████| 469/469 [00:20<00:00, 22.84it/s]



Test set: Average loss: 0.0273, Test Accuracy: 9920/10000 (99.20%)



## Net 6 training and analysis
### Results
- Best training accuracy  - 99.12
- Best test accuracy - 99.02
## Analysis 
- Model does not show overfit as much as other models
- but still does not show high enough accuracy to be included in further analysis

In [ ]:
model = Net6().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)
for epoch in range(1, 21):
    print(f"Epoch {epoch}")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

Epoch 1


Loss=0.16448640823364258 Batch_id=468 Train Accuracy=70.29: 100%|██████████| 469/469 [00:20<00:00, 23.10it/s]



Test set: Average loss: 0.1825, Test Accuracy: 9474/10000 (94.74%)

Epoch 2


Loss=0.15098947286605835 Batch_id=468 Train Accuracy=96.56: 100%|██████████| 469/469 [00:20<00:00, 22.68it/s]



Test set: Average loss: 0.1017, Test Accuracy: 9703/10000 (97.03%)

Epoch 3


Loss=0.032509945333004 Batch_id=468 Train Accuracy=97.54: 100%|██████████| 469/469 [00:20<00:00, 22.96it/s]



Test set: Average loss: 0.0627, Test Accuracy: 9819/10000 (98.19%)

Epoch 4


Loss=0.09483862668275833 Batch_id=468 Train Accuracy=98.06: 100%|██████████| 469/469 [00:20<00:00, 23.02it/s]



Test set: Average loss: 0.0699, Test Accuracy: 9782/10000 (97.82%)

Epoch 5


Loss=0.05638976767659187 Batch_id=468 Train Accuracy=98.43: 100%|██████████| 469/469 [00:20<00:00, 23.10it/s]



Test set: Average loss: 0.0518, Test Accuracy: 9828/10000 (98.28%)

Epoch 6


Loss=0.04528496041893959 Batch_id=468 Train Accuracy=98.58: 100%|██████████| 469/469 [00:20<00:00, 23.19it/s]



Test set: Average loss: 0.0457, Test Accuracy: 9863/10000 (98.63%)

Epoch 7


Loss=0.057056963443756104 Batch_id=468 Train Accuracy=98.67: 100%|██████████| 469/469 [00:20<00:00, 23.21it/s]



Test set: Average loss: 0.0370, Test Accuracy: 9890/10000 (98.90%)

Epoch 8


Loss=0.008351652882993221 Batch_id=468 Train Accuracy=98.78: 100%|██████████| 469/469 [00:20<00:00, 23.04it/s]



Test set: Average loss: 0.0343, Test Accuracy: 9885/10000 (98.85%)

Epoch 9


Loss=0.02155214548110962 Batch_id=468 Train Accuracy=98.88: 100%|██████████| 469/469 [00:20<00:00, 22.78it/s]



Test set: Average loss: 0.0321, Test Accuracy: 9889/10000 (98.89%)

Epoch 10


Loss=0.03962524235248566 Batch_id=468 Train Accuracy=99.00: 100%|██████████| 469/469 [00:20<00:00, 22.96it/s]



Test set: Average loss: 0.0315, Test Accuracy: 9900/10000 (99.00%)

Epoch 11


Loss=0.008321733213961124 Batch_id=468 Train Accuracy=99.05: 100%|██████████| 469/469 [00:20<00:00, 22.79it/s]



Test set: Average loss: 0.0301, Test Accuracy: 9900/10000 (99.00%)

Epoch 12


Loss=0.018076619133353233 Batch_id=468 Train Accuracy=99.08: 100%|██████████| 469/469 [00:20<00:00, 22.77it/s]



Test set: Average loss: 0.0277, Test Accuracy: 9906/10000 (99.06%)

Epoch 13


Loss=0.030541161075234413 Batch_id=468 Train Accuracy=99.14: 100%|██████████| 469/469 [00:20<00:00, 22.92it/s]



Test set: Average loss: 0.0282, Test Accuracy: 9905/10000 (99.05%)

Epoch 14


Loss=0.03334817662835121 Batch_id=468 Train Accuracy=99.16: 100%|██████████| 469/469 [00:20<00:00, 22.63it/s]



Test set: Average loss: 0.0281, Test Accuracy: 9912/10000 (99.12%)

Epoch 15


Loss=0.007594197988510132 Batch_id=468 Train Accuracy=99.20: 100%|██████████| 469/469 [00:20<00:00, 22.61it/s]



Test set: Average loss: 0.0307, Test Accuracy: 9902/10000 (99.02%)

Epoch 16


Loss=0.009184148162603378 Batch_id=468 Train Accuracy=99.18: 100%|██████████| 469/469 [00:20<00:00, 22.79it/s]



Test set: Average loss: 0.0278, Test Accuracy: 9912/10000 (99.12%)

Epoch 17


Loss=0.013842704705893993 Batch_id=468 Train Accuracy=99.25: 100%|██████████| 469/469 [00:20<00:00, 22.66it/s]



Test set: Average loss: 0.0258, Test Accuracy: 9917/10000 (99.17%)

Epoch 18


Loss=0.048769351094961166 Batch_id=468 Train Accuracy=99.29: 100%|██████████| 469/469 [00:20<00:00, 22.95it/s]



Test set: Average loss: 0.0318, Test Accuracy: 9896/10000 (98.96%)

Epoch 19


Loss=0.01981588639318943 Batch_id=468 Train Accuracy=99.34: 100%|██████████| 469/469 [00:20<00:00, 22.89it/s]



Test set: Average loss: 0.0269, Test Accuracy: 9917/10000 (99.17%)

Epoch 20


Loss=0.020246291533112526 Batch_id=468 Train Accuracy=99.33: 100%|██████████| 469/469 [00:20<00:00, 22.87it/s]



Test set: Average loss: 0.0259, Test Accuracy: 9906/10000 (99.06%)

